### ISSUE TITLE
Allow different entry-points rather than main.dart

### ISSUE THREAD

**Issue by lesnitsky** [invertase/melos#164]
[This](https://github.com/invertase/melos/blob/master/packages/melos/lib/src/package.dart#L744) condition is not really valid. You can have different file names and launch your app via `flutter run -t my_entry_point.dart`. I think a better way to determine that melos is running inside the app context is analyzing native folders structure

**Comment by LaysDragon** [comment]
yeah, we definitely need this,don't even mention flutter can have flavors with different entry point for different building scenario.

**Comment by blaugold** [comment]
I think we could account for multiple entry points by looking at all files directly under `/lib` and see if it contains a `main` function.

That would be simpler than analyzing the platform folders, but would not work if none of the entry points is under `/lib` (not sure that that is common).

**Comment by fnicastri** [comment]
What about an entry in the melos.yaml?
entrypoint + some others conf like the flavor.

Use case:

I have an app with 4 entry points and 3 flavors.

I have specified the run configurations manually for 3 of them (dev,staging,prduction) but I have to fix the main.dart one every time because melos don't know that I need to specify a flavor.
Would be nice to have this.

Thanks

**Comment by LaysDragon** [comment]
> What about an entry in the melos.yaml? entrypoint + some others conf like the flavor.
> 
> Use case:
> 
> I have an app with 4 entry points and 3 flavors.
> 
> I have specified the run configurations manually for 3 of them (dev,staging,prduction) but I have to fix the main.dart one every time because melos don't know that I need to specify a flavor. Would be nice to have this.
> 
> Thanks

Configurable entry point would be really handy.I definitely have to write extra intellij idea run configuration to replace the auto generated one because of the flavor problem.

**Comment by fnicastri** [comment]
@LaysDragon 
If you define your run conf Melos will not delete them, only downside is that those conf are not autogenerated and not updated when you bootstrap the workspace. 

it would be great to have entries in the Melos yaml conf by the way

**Comment by maximveksler** [comment]
Kindly consider adding support for searching for flavored apps under /lib.

I've hit this issue using `very_good_cli` starter tempate.

#### Reproductions steps

1. Generate the project `very_good create --desc 'Reframed Data Population' --org-name 'reframed.app' -t core data_populate`
2. See the listing under `ls lib/*`

```
m@m-inspiron5567:~/code/reframed/app/apps/data_populate$ ls -l lib 
total 28
drwxrwxr-x 3 m m 4096 Dec  3 22:12 app
-rw-rw-r-- 1 m m  897 Dec  3 22:12 bootstrap.dart
drwxrwxr-x 4 m m 4096 Dec  3 22:12 counter
drwxrwxr-x 3 m m 4096 Dec  3 22:12 l10n
-rw-rw-r-- 1 m m  141 Dec  3 22:12 main_development.dart
-rw-rw-r-- 1 m m  141 Dec  3 22:12 main_production.dart
-rw-rw-r-- 1 m m  141 Dec  3 22:12 main_staging.dart
```

**Comment by maximveksler** [comment]
Or you could take the "executable markdown" approach, similarly to how tea does it https://tea.xyz/tea-cli/ and just figure our the main and flavorus from the generated markdown file itself.

**Comment by fnicastri** [comment]
> > Kindly consider adding support for searching for flavored apps under /lib.
> 
> Or you could take the "executable markdown" approach, similarly to how tea does it https://tea.xyz/tea-cli/ and just figure our the main and flavorus from the generated markdown file itself.

Please don’t.

**Comment by jvincek** [comment]
+1 for this. The check `bool get isFlutterApp` in [package.dart](https://github.com/invertase/melos/blob/main/packages/melos/lib/src/package.dart) is insufficient in it's current form and could produce a wrong value if the project uses flavors and removes the standard `main.dart` (which many projects do in my experience).

While IDE run configurations can be easily fixed, the bigger issue is, are there any other features of melos (current and future) that depend on `isFlutterApp` producing a reliable/correct value? Could they potentially not work correctly for flutter apps if `isFlutterApp` returns false?

My current workaround (the least "bad" thing I came up with) is to leave `main.dart` there (for melos to detect, since it just checks if the file exists). It does not have the the normal `main()` so it can't be run directly (accidentally), but instead has `runMain(FlavorConfig config)`. And then "real" main files (e.g. `main_dev.dart` and `main_prod.dart`) call `runMain(...)` from `main.dart`. Melos will indeed produce an invalid run configuration but I can live with that, for me it's more important to have peace of mind knowing that melos can detect this is a flutter app :)

Something like this:
```dart
// flavors.dart
class FlavorConfig {
  // whatever you need to configure per-flavor
}
```

```dart
// main.dart - leave original filename for melos to detect, even though it does not have "main()"
void runMain(FlavorConfig config) {
  // Sets up everything (WidgetsFlutterBinding, Firebase...) and then calls runApp()
}
```

```dart
// main_dev.dart - only calls runMain with the flavor configuration, does nothing else (DRY)
import 'main.dart';
void main() {
  runMain(
    FlavorConfig(/* ... */),
  );
}

```

**Comment by spydon** [comment]
This was finally fixed in v8.9.0!
